torch/autograd/function.py

```diff
@@ -372,3 +381,71 @@ def forward_extended(self, *input):
 
     def backward_extended(self, *grad_output):
         raise NotImplementedError
+
+
+def symbolic_override(symbolic_fn):
+    """
+    Decorator to override ONNX export of the a given Python function with
+    specified subgraph.
+
+    Effectively allows to attach symbolic() implementation to an arbitrary
+    python function. Requirements to the decorated function:
+     - being non-member function
+     - positional inputs are Variables/Tensors or (nested) lists or tuples of
+       them (similar requirement to NestedIOFunction)
+     - outputs are similarly Variables/Tensors or (nested) lists or tuples of
+       them
+     - keyword arguments are of non-tensor type
+     - note, that function does NOT have to be autograd.Function
+
+    Example usage:
+
+    ```
+    def symb(g, x, y):
+        return g.op('Sum', x, y[0], y[1])
+
+    @symbolic_override(symb)
+    def foo(x, y):
+        return x + y[0] + y[1]
+    ```
+    """
+
+    def wrapper_maker(fn):
+
+        def wrapper(*args, **kwargs):
+            output = fn(*args, **kwargs)
+            flat_args = tuple(_iter_variables(args))
+            if not any(map(_C._jit_is_tracing, flat_args)):
```

Comment:
This flattening is quite expensive and will probably have a perf impact on regular non-ONNX, non-traced codepath. I wrote the code as it was previously in a way that makes this check cheap on purpose.

Reply:
Is it in general true that all variables would be traced if tracing is on? Or is it possible that only a subset would be?

Reply:
In general I think that would be a reasonable assumption. If we don't know some Variables we insert Constant nodes, but I don't think you can export these to ONNX (at least you probably shouldn't be by default, because it's likely that we screwed up the tracing somehow).

BTW, about reducing the cost of these checks. I realized that [we actually have a global flag](https://github.com/pytorch/pytorch/blob/master/torch/jit/__init__.py#L22-L25) that is on only if any thread (likely this one) is tracing. Checking this before you start flattening is enough to ensure we don't take a hit on perf in the regular case.